### PR TITLE
Implement hiprtc API for basic runtime compilation cases

### PR DESCRIPTION
**Pull request by linehill**
Implement hiprtc API for basic runtime compilation cases

Supported:

* Invoking extern "C" `__global__` kernels.

Unsupported or missing features:

* Compile options (currently ignored).
* Invoking non-extern "C" `__global__` kernels (required `hiprtcAddNameExpression()` and `hiprtcGetLoweredName()` functions are not implemented yet).

### CODE REVIEW

**Review by pjaaskel — APPROVED** [review]
Great work, Henry. Does this enable some tests to the test suite?

**Comment by linehill** [comment]
> Does this enable some tests to the test suite?

A test is enabled by this commit: 16b2712.

**Comment by pjaaskel** [comment]
@pvelesko is this OK to push? Does it suffice for libceed?

**Comment by pvelesko** [comment]
I will review today or tomorrow - working on HIP submodule refactoring that's required for CHIP-SPV integration into HIPCOMMON.

**Comment by pvelesko** [comment]
```
/home/pvelesko/space/CHIP-SPV/src/Utils.cc:27:10: fatal error: 'filesystem' file not found
```

I need to replace this with `experimental::filesystem` to get past compilation. @linehill

**Comment by linehill** [comment]
<filesystem> should be provided by C++17 standard. AFAIK, we don't require the standard but the root CMakeLists.txt defaults to it:
```
set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard to use.")
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

**Comment by pvelesko** [comment]
Can you push the fix? @linehill

I'm in the process of replacing fs with experimental fs but still running into linking issues

**Comment by pvelesko** [comment]
Not sure what you mean by that? I'm using the unmodified CMakeLists.txt

**Comment by linehill** [comment]
<filesystem> is part of C++17. Assuming CMake compiles the CHIP-SPV with c++17 the header should be found. Could you investigate why it doesn't find it in your system?

**Comment by pvelesko** [comment]
> Could you investigate why it doesn't find it in your system?

Ok I'll look into that.

**Comment by linehill** [comment]
Could be caused by old libstdc++ which lacks the header. Clang does not provide it and uses what is found in the system instead, I think.

**Comment by pvelesko** [comment]
Can we merge the fix then? HIPCC uses std::experimental::fs as well so this wouldn't introduce anything new.

**Comment by linehill** [comment]
The fix may fail on newer standard library not having the <experimental/filesystem> header anymore. HIPCC also uses <filesystem> if the system provides it. I'm working on an alternate fix candidate.

File: src/CHIPBackend.cc
Comment (pvelesko) [inline src/CHIPBackend.cc:1460]: thread safe?

File: src/CHIPBackend.hh
Comment (pvelesko) [inline src/CHIPBackend.hh:1891]: thread safe?

File: include/hip/spirv_hip_fp16.h
Comment (pvelesko) [inline include/hip/spirv_hip_fp16.h:43]: I was confused about this part.. Where does api_half come from?

File: src/CHIPBackend.cc
Comment (pvelesko) [inline src/CHIPBackend.cc:741]: thread safe?

File: include/hip/spirv_hip_fp16.h
Comment (linehill) [reply include/hip/spirv_hip_fp16.h:43]: No idea where it was supposed to come from but it not defined anywhere, AFAIK.

File: src/CHIPBackend.cc
Comment (linehill) [reply src/CHIPBackend.cc:1460]: I have a follow up PR for removing this. I'll transfer it here.

File: src/CHIPBackend.hh
Comment (linehill) [reply src/CHIPBackend.hh:1891]: I'll remove this.

File: src/Utils.hh
Comment (pvelesko) [inline src/Utils.hh:32]: unnecessary if included in common.hh

**Comment by pvelesko** [comment]
Nice work! I think we should merge #182 first as that will enable a lot more hipRTC tests.

**Comment by linehill** [comment]
I'll rebase this patch when it's merged.

File: src/CHIPBackend.cc
Comment (linehill) [reply src/CHIPBackend.cc:742]: Added device scope lock.

File: src/Utils.hh
Comment (linehill) [reply src/Utils.hh]: Filesystem include is gone. Didn't see optional included in common.h.

**Comment by linehill** [comment]
Updated. The patch tries to use `<experimental/filesystem>` header instead if the `<filesystem>` header is not available.

**Comment by pvelesko** [comment]
Merged the HIPCOMMON rebase. Please checkout HIP/test/README there's a new section about hipRTC testing

**Comment by pvelesko** [comment]
Still waiting on feedback from AMD and the rebased branch needs some additional work so go ahead and merge this when you are ready.

**Comment by linehill** [comment]
Alright, thanks.

**Pull request merged**
